Adding optional volume support to RDL

cp16net opened the pull request:
- Allowing a user to add a volume to the instance on create api call
- added some helper code to make sure the mysql service is up 
- add a config value to turn on volume support
- moving around the lines to be dependent on volume support
- trying to stop volume support
- fixes with the merge
- imported and raised exceptions in guest correctly
- fixed the config get values as boolean and not just the string
- deleting the serivce status entry on delete of the instance so that the
  guest can not update the instance to active again.
- made the create and update times saved correctly for an instance
- check if mysql is already installed in the guest
- clean up instance model
- using _(i8n)
- adding the populate databases back in
- pep8 compliance

**after comments**
- updated apt get version method
- moved the volumeHelper class to a new module
- made the start/stop mysql methods public
- created a method to create volume for instances and readability
- created a new client for volume calls
- added the volume and size to the create/list/show instance api calls

cp16net committed:
Allowing a user to add a volume to the instance on create api call

added some helper code to make sure the mysql service is up

add a config value to turn on volume support

moving around the lines to be dependent on volume support

trying to stop volume support

fixes with the merge

imported and raised exceptions in guest correctly

fixed the config get values as boolean and not just the string

deleting the serivce status entry on delete of the instance so that the

guest can not update the instance to active again.

made the create and update times saved correctly for an instance

check if mysql is already installed in the guest

clean up instance model

using _(i8n)

adding the populate databases back in

pep8 compliance

updated apt get version method

moved the volumeHelper class to a new module

made the start/stop mysql methods public

created a method to create volume for instances and readability

created a new client for volume calls

added the volume and size to the create/list/show instance api calls

removing the mount_point and device_path from the MysqlApp class

removing code that deleted the instance status entry

adding volume service to the reddwarf.conf.test file

set volumes to not be disabled by default

TimSimpsonR commented on reddwarf/common/remote.py:
```diff
@@ -40,8 +39,13 @@ def create_nova_client(context):
-    SERVICE_TYPE = CONFIG.get('nova_service_type', 'compute')
-    SERVICE_NAME = CONFIG.get('nova_service_name', 'Compute Service')
+
+    if service_type and service_type.lower() == "volume":
```
Instead of having this be an argument, how about create another method to "create_volume_client" and one more called "create_compute_client" that called this one and potentially sent in the SERVICE_TYPE and SERVICE_NAME variables as well (so the volume and compute specific code would stay in those two respective methods).

cp16net replied:
ok i split this into 2 clients compute/volume

TimSimpsonR commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -593,6 +605,38 @@ def install_and_secure(self, pkg, memory_mb):
 
+    def _is_mysql_installed(self):
+        #TODO(cp16net) need to perform check if mysql is installed
+        return True
```
We already have MySqlAppStatus.is_mysql_installed.

cp16net replied:
good point

TimSimpsonR commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -736,8 +779,90 @@ def start_mysl_with_conf_changes(self, pkg, updated_memory_mb):
         self._internal_stop_mysql(True)
+
+
+class VolumeHelper(object):
```
This is important enough to get its own Python module.

cp16net replied:
thats a good idea since it is independent.

TimSimpsonR commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -488,13 +492,16 @@ def enable_root(self):
-        app = MySqlApp(self.status)
+        app = MySqlApp(self.status, device_path, mount_point)
+        #TODO(cp16net) how to handle mounting the volume for migration
+        app.mount_volume()
```
I don't like that "mount_volume" lives in MySqlApp, a class that exists just to manage MySQL. It should be in its own class just for working with the volume stuff.

To mount the volume in prepare, you'd simply make an instance of this new class, call mount_volume, then make an instance of MySqlApp and call install_and_secure like before.

TimSimpsonR commented on reddwarf/guestagent/pkg.py:
```diff
@@ -157,7 +157,8 @@ def _remove(self, package_name, time_out):
         try:
-            utils.execute("apt-get", "update", run_as_root=True)
+            #(cp16net) the run_as_root doesnt work correctly so i added sudo
+            utils.execute("sudo", "apt-get", "update", run_as_root=True)
```
Good catch. :)

hub-cap replied:
This should not have sudo in it. It should instead do this.

utils.execute(*args, run_as_root=True, root_helper="sudo")

cp16net replied:
oh thats nice.

hub-cap replied:
Ya thats how the openstack common helper class works. In the guest code im writing (which i will one day check in), ive created a _exec() so u dont ahve to pass those, but u can optionally pass root=True|False to it

TimSimpsonR commented on reddwarf/instance/models.py:
```diff
@@ -128,6 +129,8 @@ def delete(self, force=False):
         self.db_info.task_status = InstanceTasks.DELETING
         self.db_info.save()
+        # Delete the service_statuses entry
+        self.service_status.delete()
```
This is unnecessary.  The guest should be allowed to do this as the service status must always remain a source of truth as to what the service is up to on the guest at any time, even if its about to be deleted, so we can see this in the management API and to keep the code simple. If the REST API status needs to stay in the DELETED state regardless of what the guest says then the task status along with the code for the Instance.status property can be used to accomplish that.

TimSimpsonR commented on reddwarf/instance/models.py:
```diff
@@ -140,21 +143,74 @@ def _delete_server(self):
             task_status=InstanceTasks.NONE)
         LOG.debug(_("Created new Reddwarf instance %s...") % db_info.id)
+
+        volume_support = config.Config.get("reddwarf_volume_support", 'False')
```
If everything from this line to the end of the else block was moved to a new method or free-standing function it would be easier to read.

cp16net replied:
agree

TimSimpsonR commented on reddwarf/instance/models.py:
```diff
@@ -346,10 +405,26 @@ def load(context):
+            except ModelNotFoundError:
+                LOG.info(_("Status entry not found either failed to start "
+                           "or instance was deleted"))
+                continue
```
Very nice.

TimSimpsonR commented on reddwarf/tests/fakes/guestagent.py:
```diff
@@ -68,6 +68,7 @@ def list_users(self):
     def prepare(self, memory_mb, databases, users):
```
If this isn't changed to allow more parameters fake mode won't work.

cp16net replied:
`god_mode` :godmode:

hub-cap replied:
we need to kwargs it.

TimSimpsonR commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -593,6 +605,38 @@ def install_and_secure(self, pkg, memory_mb):
+            # mounting it in /var/lib/mysql and restarting mysql service
+            restart_mysql = False
+            VolumeHelper.format(self.device_path)
+            if self._is_mysql_installed():
```
I'm not sure why this is is here as right now "mount_volume" is only called by prepare. If there are other use cases there should be tests written too, and this code should be part of those pull requests.

This code can still be separated out of MySqlApp. To do it, you'd need to move the logic of this to the host side (formerly called compute manager) and have it make several guest calls, first to stop_sql, then mount the volume and do rsync, and then start_mysql.

TimSimpsonR commented:
Is there another pull request for Ridley to turn on the volume stuff?

ed- commented on reddwarf/common/utils.py:
```diff
@@ -208,10 +208,11 @@ def get_id_from_href(href):
-           + str(kwargs))
+        raise exception.ProcessExecutionError("Time out after waiting "
+            + str(time) + " seconds when running proc: " + str(args)
+            + str(kwargs))
```
Grumble grumble format strings over concatenation

Alternately you know that 
`("string 1"
  "string 2"
  "string 3")`
is concatenated for free right

rnirmal commented on reddwarf/instance/models.py:
```diff
@@ -140,21 +143,74 @@ def _delete_server(self):
+                # <id>:[<type>]:[<size(GB)>]:[<delete_on_terminate>]
+                # setting the delete_on_terminate instance to true=1
+                mapping = "%s:%s:%s:%s" % (v_ref.id, '', v_ref.size, 1)
+                block_device = {'vdb': mapping}
```
This mapping device needs to be configurable. and we may have to do a little more trickery here. We don't know what's the next device available on the guest. Also in cases for ovz where this is mounted on the host, that's not going to work for us... Sorry didn't point out this earlier, just wanted to get that one volume working I guess :)

cp16net replied:
Yeah after we talked i understand there are some issues with ovz here but not sure how to handle that yet.

TimSimpsonR commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -536,12 +560,15 @@ class MySqlApp(object):
+            'state_change_wait_time', 2 * 60))
         self.status = status
+        self.device_path = device_path
+        self.mount_point = mount_point
```
"device_path" and "mount_point" can be removed from this class.

cp16net replied:
removed.

TimSimpsonR commented on reddwarf/guestagent/pkg.py:
```diff
@@ -171,44 +173,45 @@ def pkg_install(self, package_name, time_out):
+                    return parts[2]
+        msg = _("version() saw unexpected output from dpkg!")
+        LOG.error(msg)
+        raise exception.GuestError(msg)
```
Was there a reason this was rewritten to not use pexpect? I know we've had problems with the pkg module on RDL, but the cowsay tests were running successfully against this for a year, so I'm surprised it was that bad.

TimSimpsonR replied:
Since there was a pexpect bug I guess this is what we have to live with... geeze... :(

cp16net replied:
Yeah i didnt know what else to do because the columns from pexpect were being truncated to 8 characters. No idea why this was caused but this was the work around.

hub-cap commented on reddwarf/common/remote.py:
```diff
@@ -55,10 +55,38 @@ def create_nova_client(context):
+    SERVICE_TYPE = CONFIG.get('nova_volume_service_type', 'volume')
+    SERVICE_NAME = CONFIG.get('nova_volume_service_name', 'Volume Service')
+
+    #TODO(cp16net) need to fix this proxy_tenant_id
```
Whats this comment mean exactly?

cp16net replied:
This is from the copy pasta dinner we had for the `create_nova_client` above. I believe that was to make sure we added it to the context.

hub-cap replied:
mmmmm copy pasta!

hub-cap commented on reddwarf/instance/service.py:
```diff
@@ -56,7 +56,10 @@ class BaseController(wsgi.Controller):
     def __init__(self):
-        self.add_addresses = config.Config.get('add_addresses', False)
+        self.add_addresses = utils.bool_from_string(
+                        config.Config.get('add_addresses', 'False'))
```
How come these are coming back as strings?

cp16net replied:
great question! well because they are all strings in the configuration file and no types are associated with them from the config class. 

I realized this when i attempted to turn off volume support with False. It was still turned on. because the string existed. I figured out this is what melange was doing to convert the values. Integers should be casted as well when pulling a config value.

hub-cap replied:
Ugh wonderful :)

Sent from my digital shackles. 

On Apr 26, 2012, at 1:51 PM, Craigreply@reply.github.com wrote:

> > @@ -56,7 +56,10 @@ class BaseController(wsgi.Controller):
> >         }
> > 
> > ```
> > def __init__(self):
> > ```
> > -        self.add_addresses = config.Config.get('add_addresses', False)
> > -        self.add_addresses = utils.bool_from_string(
> > -                        config.Config.get('add_addresses', 'False'))
> 
> great question! well because they are all strings in the configuration file and no types are associated with them from the config class. 
> 
> I realized this when i attempted to turn off volume support with False. It was still turned on. because the string existed. I figured out this is what melange was doing to convert the values. Integers should be casted as well when pulling a config value.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/hub-cap/reddwarf_lite/pull/65/files#r740775

hub-cap commented:
just to make sure, plz rerun pep8 after you do all the changes :D

cp16net commented:
updated. the pep8 gods were happy.

ed- commented on reddwarf/common/utils.py:
```diff
@@ -208,10 +210,12 @@ def get_id_from_href(href):
-           + str(kwargs))
+        msg = _("Time out after waiting"
+                " %(time)s seconds when running proc: %(args)s"
+                " %(kwargs)s") % locals()
```
golf clap

ed- commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -176,8 +181,8 @@ def _get_actual_db_status(self):
-                out, err = utils.execute_with_timeout("/bin/ps", "-C", "mysqld",
-                                                      "h")
+                out, err = utils.execute_with_timeout("/bin/ps", "-C",
+                                                      "mysqld", "h")
```
Do you think this would be any easier to read/maintain?

```
cmd = "/bin/ps -C mysqld h"
cmd = cmd.split(' ')
out, err = utils.execute_with_timeout(*cmd)
```

No holdup, just a suggestion to keep you from having to wrap these lines every time you call it

rnirmal replied:
that's a good suggestion, but we can punt on it for now.. too many changes already :)

ed- commented:
:godmode: +1

rnirmal commented on reddwarf/guestagent/dbaas.py:
```diff
@@ -488,13 +492,33 @@ def enable_root(self):
+        # status end_mysql_install set with install_and_secure()
         app = MySqlApp(self.status)
+        restart_mysql = False
+        if not device_path is None:
```
This could just be `if device_path:`

rnirmal commented on reddwarf/instance/models.py:
```diff
@@ -45,12 +46,40 @@ def load_server(context, instance_id, server_id):
+        volume_client = create_nova_volume_client(context)
+        try:
+            volumes = []
+            if utils.bool_from_string(volume_support):
```
Don't think that's needed. You are already in a if block for volume_support.

rnirmal commented on reddwarf/instance/models.py:
```diff
@@ -140,25 +172,93 @@ def _delete_server(self):
+        if utils.bool_from_string(volume_support):
+            LOG.debug(_("Starting to create the volume for the instance"))
+            volume_client = create_nova_volume_client(context)
+            volume_desc = ("mysql volume for %s" % context.tenant)
```
can you add `db_info.id` instead of `context.tenant` for the description. Then we could just use that description to pass on the the backend systems

cp16net committed:
clean up and change the volume desription to have the instance id

fake mode works with all the tests

mounting a volume is causing a database name of #mysql50#lost+found to be created

pep8 fixes

cp16net commented:
working on getting fake mode working now.

cp16net commented:
Fake mode and real mode working for tests.

TimSimpsonR commented:
Ok, after getting it explained to me I guess we just have two versions of MySQL to go before this stops being an issue. That's ok, for now.
+1

rnirmal commented:
+1.. .needs rebasing

cp16net committed:
pep8 fixes and resolve a few other rebase issues

cp16net commented:
rebased and passed everything including PEP8 :godmode:
